### ISSUE TITLE
Exporting RTDETR model to ONNX fails

### ISSUE THREAD

**Issue by Maxwell12345** [ultralytics/ultralytics#5299]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

from ultralytics import RTDETR

model = RTDETR('rtdetr-l.pt')

model.info()

model.export(format="onnx")
and
model.export(format="onnx", dynamic=True)

both give the same trace stack
PyTorch: starting from rtdetr-l.pt with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 1, 300, 4), (1, 1, 300, 80)) (63.4 MB)

ONNX: starting export with onnx 1.13.0 opset 17...
ONNX: export failure ❌ 1.6s: Unsupported: ONNX export of operator get_pool_ceil_padding, input size not accessible. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues  [Caused by the value 'input.4 defined in (%input.4 : Float(*, *, *, *, strides=[3297312, 103041, 321, 1], requires_grad=0, device=cpu) = onnx::Pad[mode="constant"](%452, %479, %455), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /tf/ultralytics-vit/ultralytics/nn/modules/block.py:71:0
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Pad'.] 
    (node defined in /tf/ultralytics-vit/ultralytics/nn/modules/block.py(71): forward
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1508): _slow_forward
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/tf/ultralytics-vit/ultralytics/nn/tasks.py(64): _forward_once
/tf/ultralytics-vit/ultralytics/nn/tasks.py(215): forward
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1508): _slow_forward
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/usr/local/lib/python3.11/dist-packages/torch/jit/_trace.py(124): wrapper
/usr/local/lib/python3.11/dist-packages/torch/jit/_trace.py(133): forward
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.11/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/usr/local/lib/python3.11/dist-packages/torch/jit/_trace.py(1285): _get_trace_graph
/usr/local/lib/python3.11/dist-packages/torch/onnx/utils.py(915): _trace_and_get_graph_from_model
/usr/local/lib/python3.11/dist-packages/torch/onnx/utils.py(1011): _create_jit_graph
/usr/local/lib/python3.11/dist-packages/torch/onnx/utils.py(1135): _model_to_graph
/usr/local/lib/python3.11/dist-packages/torch/onnx/utils.py(1596): _export
/usr/local/lib/python3.11/dist-packages/torch/onnx/utils.py(516): export
/tf/ultralytics-vit/ultralytics/yolo/engine/exporter.py(314): export_onnx
/tf/ultralytics-vit/ultralytics/yolo/engine/exporter.py(112): outer_func
/tf/ultralytics-vit/ultralytics/yolo/engine/exporter.py(237): __call__
/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py(115): decorate_context
/tf/ultralytics-vit/ultralytics/vit/rtdetr/model.py(112): export
/usr/local/lib/python3.11/dist-packages/torch/utils/_contextlib.py(115): decorate_context
/tmp/ipykernel_4114/936804572.py(6): <module>
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py(3526): run_code
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py(3466): run_ast_nodes
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py(3284): run_cell_async
/usr/local/lib/python3.11/dist-packages/IPython/core/async_helpers.py(129): _pseudo_sync_runner
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py(3079): _run_cell
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py(3024): run_cell
/usr/local/lib/python3.11/dist-packages/ipykernel/zmqshell.py(546): run_cell
/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py(422): do_execute
/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py(740): execute_request
/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py(412): dispatch_shell
/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py(505): process_one
/usr/local/lib/python3.11/dist-packages/ipykernel/kernelbase.py(516): dispatch_queue
/usr/lib/python3.11/asyncio/events.py(80): _run
/usr/lib/python3.11/asyncio/base_events.py(1909): _run_once
/usr/lib/python3.11/asyncio/base_events.py(604): run_forever
/usr/local/lib/python3.11/dist-packages/tornado/platform/asyncio.py(195): start
/usr/local/lib/python3.11/dist-packages/ipykernel/kernelapp.py(736): start
/usr/local/lib/python3.11/dist-packages/traitlets/config/application.py(1046): launch_instance
/usr/local/lib/python3.11/dist-packages/ipykernel_launcher.py(17): <module>
<frozen runpy>(88): _run_code
<frozen runpy>(198): _run_module_as_main
)

    Inputs:
        #0: 452 defined in (%452 : Float(1, 32, 320, 320, strides=[3276800, 102400, 320, 1], requires_grad=0, device=cpu) = onnx::Relu(%input), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0/ultralytics.nn.modules.conv.Conv::stem1/torch.nn.modules.activation.ReLU::act # /usr/local/lib/python3.11/dist-packages/torch/nn/functional.py:1469:0
    )  (type 'Tensor')
        #1: 479 defined in (%479 : Long(8, strides=[1], device=cpu) = onnx::Cast[to=7](%478), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /tf/ultralytics-vit/ultralytics/nn/modules/block.py:71:0
    )  (type 'Tensor')
        #2: 455 defined in (%455 : NoneType = prim::Constant(), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0
    )  (type 'NoneType')
    Outputs:
        #0: input.4 defined in (%input.4 : Float(*, *, *, *, strides=[3297312, 103041, 321, 1], requires_grad=0, device=cpu) = onnx::Pad[mode="constant"](%452, %479, %455), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /tf/ultralytics-vit/ultralytics/nn/modules/block.py:71:0
    )  (type 'Tensor')

### Environment

docker ran jupyter notebook

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Maxwell12345 hello,

Thank you for bringing this issue to our attention. I see that there seems to be an issue with exporting the RTDETR model to ONNX format. The error message you've shared is indicating that the ONNX export of the operator 'get_pool_ceil_padding' isn't supported, as the input size isn't accessible.

Unfortunately, not all PyTorch functionalities are supported in ONNX due to the difference in their internal platforms and this might just be an instance of that. PyTorch uses dynamic computation graphs, which is a powerful feature allowing changes in the way operations are performed based on the input data. On the other hand, ONNX uses static computation graphs where everything needs to be determined beforehand, suitable for deployment in less flexible hardware.

The operator 'get_pool_ceil_padding', which seems to be involved in this issue, seems more PyTorch specific and could be one reason for the issue.

We'll need to investigate this issue further to identify any possible workarounds or solutions. This could involve implementing a custom function to replace the unsupported operation or changing the model architecture. Given the nature of the issue, it could take some time to resolve.

We'll aim to give you an update as soon as there's any progress on this. In the meantime, please feel free to report any other issues or questions you have.

Best Regards

**Comment by Maxwell12345** [comment]
Hey! Thank you for the response, sir. This is invaluable information. I will see if I can implement it into the code my self and I will be sure to share if I find a solution. I love this and other open source ML communities so anywhere I can give code is always first in mind.

Unrelated to the issue, why does Ultralytics opt for using PyTorch over Pickle, Keras, TF, etc?

We can take this offline if it's not the right place but I do have some questions as I'm new to computer vision (experience is mostly LLMs and GANs)

**Comment by glenn-jocher** [comment]
@Maxwell12345 hello! I'm really glad to hear that you found the information helpful and we'd be thrilled if you'd share any solutions you find. We greatly appreciate your interest in contributing to the community.

Your question about why we opted for PyTorch is a great one. In essence, PyTorch provides a strong blend of flexibility and performance. Its imperative structure allows for dynamic computational graphs, which can be advantageous in projects that require flexibility, such as research prototypes or projects with variable-length inputs. We also find that PyTorch’s syntax and style is intuitive and its data loading and augmentation capabilities quite powerful. 

However, choosing a deep learning framework can be a very context-specific decision and each framework has its own strengths and weaknesses. TensorFlow, Keras, and other frameworks you mentioned all have their own respective merits and situations where they might shine.

As you continue to learn about computer vision, you'll discover that often the best tool for the job is a matter of preference or project-specific needs. I encourage you to explore different options and continue to ask insightful questions on your learning journey. Don't hesitate to bring any other questions you have, we're here to help!

**Comment by MarcoPrassel** [comment]
Hi @glenn-jocher,
I am also experiencing this problem, I am waiting for updates.
Best Regards.

**Comment by glenn-jocher** [comment]
@MarcoPrassel hi there, 
Thank you for reaching out and letting us know that you're experiencing a similar issue. We are currently looking into it and your patience is much appreciated. We understand the urgency and will strive to have an update or solution as soon as possible. Rest assured that our team is working diligently to resolve the issue. Thank you for your understanding, and we'll keep you posted about any progress. Best Regards.

**Comment by SatyamGaba** [comment]
any updates on this? I am facing same issue

**Comment by glenn-jocher** [comment]
@SatyamGaba it appears that you are encountering an issue when trying to export the RTDETR model to ONNX format. The error message indicates that the operation `get_pool_ceil_padding` is not supported for ONNX export, and the input size is not accessible.

This is a known limitation when exporting certain PyTorch operations to ONNX, as ONNX may not support all PyTorch operations or the way they are implemented in the model. The error message suggests that the issue is with the padding operation in the model, which is not compatible with ONNX's export requirements.

To resolve this issue, you have a few options:

1. **Modify the Model**: If you have access to the model's source code, you can try modifying the padding operation to use a supported method. This might involve changing the padding mode or the way the padding is calculated.

2. **Request Support**: As suggested in the error message, you can request support or submit a pull request on the PyTorch GitHub repository. This would involve the PyTorch developers potentially adding support for the operation in a future release.

3. **Use a Different Export Format**: If ONNX export is not critical for your use case, you can try exporting the model to a different format that might not have the same limitations.

4. **Wait for PyTorch Updates**: If the operation is not currently supported, you may need to wait for a future update of PyTorch that includes support for the operation in ONNX export.

5. **Use a Different Model**: If exporting to ONNX is critical and you cannot modify the model or wait for updates, you may need to consider using a different model that does not use unsupported operations.

It's important to note that model exporting can be complex, and compatibility issues like this are not uncommon. If you are not comfortable making changes to the model's source code, it may be best to seek assistance from someone with more experience in model exporting and PyTorch.

Lastly, if you are part of a team or organization with an enterprise license for Ultralytics software, you may have access to priority support that could assist with this issue.
